### PR TITLE
feat(membership): added functionality for swapping hs members

### LINKED ISSUE #105: Automatisk medlemskap til HS
Når du er gjort til leder av en undergruppe burde du automatisk få et medlemskap til HS, og tidligere leder burde fjernes.

### PR DESCRIPTION
**Pull request by Her0elt**
## Proposed changes
Added logic for promoting and swaping hs members when a leader of a subgroup is changed.

Issue number: closes #105 

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] CHANGELOG.md has been updated. [Guide](https://tihlde.slab.com/posts/changelog-z8hybjom)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Pull request title follows [conventional commits](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) (`type(scope): description`)
- [ ] Build (`make PR`) was run locally without failures

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: app/group/models/membership.py [hunk 1]
```diff
@@ -98,14 +101,46 @@ def delete(self, *args, **kwargs):
...
+                .first()
+            )
+        if seat_to_delete:
+            seat_to_delete.delete()
+            
+    @atomic
+    def swap_hs_seat(self, previous_seat):
+        seat_to_delete = None
+        if previous_seat:
+            print("skjnkjdsdkjsndskjndsk")
```
Comment (olros) [inline app/group/models/membership.py:117]: skjnkjdsdkjsndskjndsk

File: app/group/models/membership.py [hunk 2]
```diff
@@ -98,14 +101,46 @@ def delete(self, *args, **kwargs):
...
+
+        if group and self:
+            membership = Membership.objects.get_or_create(user=self.user, group=group)[
+                0
+            ]
+            membership.save()
+
     @atomic
     def swap_board(self):
+        print("skjnkjdsdkjsndskjndsk")
```
Comment (olros) [inline app/group/models/membership.py:138]: Må fjernes ;)

File: app/group/tests/test_membership_model.py [hunk 3]
```diff
@@ -82,3 +87,53 @@ def test_swap_leader_does_not_create_two_leaders(group):
...
+    Tests that if you demote as user from leader to member of a subgroup,
+    the user is automaticly removed from the HS group
+    """
+    group = GroupFactory(type=GroupType.SUBGROUP)
+    membership = MembershipFactory(membership_type=MembershipType.LEADER, group=group)
+    membership.membership_type = MembershipType.MEMBER
+    membership.save()
+    membership.refresh_from_db()
+    assert membership.membership_type == MembershipType.MEMBER
+    assert not len(Membership.objects.filter(user=membership.user, group=hs))
```
Comment (Eirsteir) [inline app/group/tests/test_membership_model.py:139]: Her kan vi bruke litt Django magi: 
```suggestion
    assert not membership.user.membership_set.filtergroup=hs).exists()
```

Hadde vi brukt `related_name=memberships` som argument [i User](https://github.com/TIHLDE/Lepton/blob/c4844b8fcd3fd329ca0051e5db3bfb965ffec537/app/group/models/membership.py#L17) kunne den blitt enda mer lesbar: 
`assert not membership.user.memberships.filter(group=hs).exists()`

Dette kan brukes i alle øvrige tester også hvis du ønsker det :)

File: app/group/tests/test_membership_model.py [hunk 4]
```diff
@@ -82,3 +87,53 @@ def test_swap_leader_does_not_create_two_leaders(group):
...
+    membership = MembershipFactory(membership_type=MembershipType.MEMBER, group=group)
+    membership.membership_type = MembershipType.LEADER
+    membership.save()
+    membership.refresh_from_db()
+    assert membership.membership_type == MembershipType.LEADER
+    assert Membership.objects.get(user=membership.user, group=hs)
+
+
+@pytest.mark.django_db
+def test_on_swap_creates_hs_membership_with_leader(hs):
```
Comment (Eirsteir) [inline app/group/tests/test_membership_model.py:108]: Kan du legge på en negativ test også som sjekker at brukeren _ikke_ blir lagt til i HS når gruppen er en komité?

File: app/group/models/membership.py [hunk 5]
```diff
@@ -98,14 +104,43 @@ def delete(self, *args, **kwargs):
         MembershipHistory.from_membership(self)
         return super().delete(*args, **kwargs)
 
+    def delete_hs_membership(self):
+        seat_to_delete = (
+            Membership.objects.select_for_update()
```
Comment (Eirsteir) [inline app/group/models/membership.py:109]: Nice!

**Review by Eirsteir — APPROVED** [review]
Ser bra ut det her! La til noen kommentarer til testene dine, men godt jobbet 😄

**Pull request merged**